File: client/src/api/migration.ts
/**
 * Data Migration API Client (VCL HDD -> SSD)
 */

import { apiClient } from '../lib/api';

// ========== Directory Browse ==========

export interface DirectoryEntry {
  name: string;
  path: string;
  is_mountpoint: boolean;
}

export async function browseDirectory(path: string): Promise<DirectoryEntry[]> {
  const res = await apiClient.get('/api/ssd/migration/browse', { params: { path } });
  return res.data;
}

// ========== Types ==========

export interface MigrationJobResponse {
  id: number;
  job_type: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  source_path: string;
  dest_path: string;
  total_files: number;
  processed_files: number;
  skipped_files: number;
  failed_files: number;
  total_bytes: number;
  processed_bytes: number;
  current_file: string | null;
  progress_percent: number;
  error_message: string | null;
  dry_run: boolean;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
}

export interface VCLMigrationStartRequest {
  source_path: string;
  dest_path: string;
  dry_run: boolean;
}

export interface VCLVerifyRequest {
  dest_path: string;
}

export interface VCLCleanupRequest {
  source_path: string;
  dry_run: boolean;
}

// ========== API Functions ==========

export async function startVCLMigration(data: VCLMigrationStartRequest): Promise<MigrationJobResponse> {
  const res = await apiClient.post('/api/ssd/migration/vcl/start', data);
  return res.data;
}

export async function startVCLVerify(data: VCLVerifyRequest): Promise<MigrationJobResponse> {
  const res = await apiClient.post('/api/ssd/migration/vcl/verify', data);
  return res.data;
}

export async function startVCLCleanup(data: VCLCleanupRequest): Promise<MigrationJobResponse> {
  const res = await apiClient.post('/api/ssd/migration/vcl/cleanup', data);
  return res.data;
}

export async function getMigrationJobs(limit = 20): Promise<MigrationJobResponse[]> {
  const res = await apiClient.get('/api/ssd/migration/jobs', { params: { limit } });
  return res.data;
}

export async function getMigrationJob(id: number): Promise<MigrationJobResponse> {
  const res = await apiClient.get(`/api/ssd/migration/jobs/${id}`);
  return res.data;
}

export async function cancelMigrationJob(id: number): Promise<void> {
  await apiClient.post(`/api/ssd/migration/jobs/${id}/cancel`);
}

File: backend/app/schemas/migration.py
"""Pydantic schemas for data migration API."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, field_validator


def _reject_path_traversal(v: str) -> str:
    """Reject paths containing '..' components."""
    if ".." in v:
        raise ValueError("Path must not contain '..'")
    return v


class VCLMigrationStartRequest(BaseModel):
    """Start a VCL blob migration from HDD to SSD."""
    source_path: str
    dest_path: str
    dry_run: bool = False

    @field_validator("source_path", "dest_path")
    @classmethod
    def validate_paths(cls, v: str) -> str:
        return _reject_path_traversal(v)


class VCLVerifyRequest(BaseModel):
    """Start integrity verification of migrated VCL blobs."""
    dest_path: str

    @field_validator("dest_path")
    @classmethod
    def validate_dest(cls, v: str) -> str:
        return _reject_path_traversal(v)


class VCLCleanupRequest(BaseModel):
    """Start cleanup of source blobs after verified migration."""
    source_path: str
    dry_run: bool = False

    @field_validator("source_path")
    @classmethod
    def validate_source(cls, v: str) -> str:
        return _reject_path_traversal(v)


class DirectoryEntry(BaseModel):
    """A directory entry returned by the browse endpoint."""
    name: str
    path: str
    is_mountpoint: bool


class MigrationJobResponse(BaseModel):
    """Migration job status and progress."""
    id: int
    job_type: str
    status: str
    source_path: str
    dest_path: str
    total_files: int
    processed_files: int
    skipped_files: int
    failed_files: int
    total_bytes: int
    processed_bytes: int
    current_file: Optional[str]
    progress_percent: float
    error_message: Optional[str]
    dry_run: bool
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    duration_seconds: Optional[float]
